### PR TITLE
Allow gateway to be driven directly using declarations

### PR DESCRIPTION
**Pull request by msavy**
Firstly, apologies for the noisy commit timeline -- I'm happy for a squash commit when the moment is right! Was a process of discovery.

Still at seeking feedback stage, but would be hugely appreciative of thoughts as I'm obviously not a pro with this codebase.

### Rationale 

* Rejig the order of commands so that the root is now:
  - `apiman gateway ...`
  - `apiman manager ...`

* Shares the same fundamental data model as manager; with a small tweak to policy which is non-breaking for the manager functionality.
* Declarative only for now.

* For both inbuilt and plugin policies it resolves the correct information needed to successfully interact with the Gateway API. This should substantially simplify things.
  - Plugins are resolved using the Manager libraries, so it should function near identically.

```yaml
# Simple apiman example
---
  system:
    gateways:
      - name: "test-gw"
        type: "REST"
        config:
          endpoint: "http://localhost:8080/apiman-gateway-api"
          username: "apimanager"
          password: "apiman123!"
    plugins:
      - name: TestPolicy
        groupId: "io.apiman.plugins"
        artifactId: "apiman-plugins-test-policy"
        version: "1.3.1.Final"
  org:
    name: "test"
    apis:
      - name: "example"
        version: "1.0"
        config:
          endpoint: "http://localhost:8080/services/echo"
          endpointType: "rest"
          public: true
          gateway: "test-gw"
        policies:
          - name: "CachingPolicy"
            config:
              ttl: 60
          - plugin: TestPolicy
            config:
              foo: 123
```

```
$ apiman-cli gateway apply -f /Users/msavy/tmp/sample.yaml
```

### TODO

- [ ] Review
- [ ] Tests
- [ ] Clients
- [ ] List policies?

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -69,6 +70,12 @@ dependencies {
     compile "org.modelmapper:modelmapper:$version_modelmapper"
     compile "org.apache.commons:commons-lang3:$version_commons_lang"
     compile "com.google.inject:guice:$version_guice"
+    
+    // apiman deps
+    compile "io.apiman:apiman-gateway-engine-beans:$version_apiman"
+    compile "io.apiman:apiman-common-plugin:$version_apiman"
+    compile "io.apiman:apiman-manager-api-core:$version_apiman"
+    runtime "io.apiman:apiman-distro-data:$version_apiman"
```
Comment (msavy) [inline build.gradle:78]: I was having issues with eclipse + gradle (probably should switch to intellij for good...); probably don't need this as a runtime dep.

`distro-data` contains various useful resources such as a list of inbuilt policies

File: src/main/java/io/apiman/cli/command/AbstractCommand.java [hunk 2]
```diff
@@ -16,24 +16,26 @@
 
 package io.apiman.cli.command;
 
-import com.google.common.collect.Maps;
+import static io.apiman.cli.util.AuthUtil.DEFAULT_SERVER_PASSWORD;
```
Comment (msavy) [inline src/main/java/io/apiman/cli/command/AbstractCommand.java:19]: Sorry, looks like my eclipse being annoying and reordering dependencies. If you have settings, I'll use those instead.

File: src/main/java/io/apiman/cli/core/declarative/command/ManagerApplyCommand.java [hunk 3]
```diff
@@ -1,11 +1,11 @@
 /*
- * Copyright 2016 Pete Cornish
+ * Copyright 2017 JBoss Inc
```
Comment (msavy) [inline src/main/java/io/apiman/cli/core/declarative/command/ManagerApplyCommand.java:2]: This wasn't deliberate, I'll change it back. Just a refactoring artifact.

**Comment by msavy** [comment]
Here's the gateway format documented informally:

```
# Simple apiman example
---
  system:
    gateways:
      - name: "test-gw"
        type: "REST"
        config:
          endpoint: "http://localhost:8080/apiman-gateway-api"
          username: "apimanager"
          password: "apiman123!"
    plugins:
      - name: test-plugin # friendly name
        groupId: "io.apiman.plugins"
        artifactId: "apiman-plugins-test-policy"
        version: "1.2.4.Final"
      - name: another-test-plugin
        groupId: "io.apiman.plugins"
        artifactId: "apiman-plugins-test-policy-2"
        version: "1.2.7.Final"

  org:
    name: "test"
    apis:
      - name: "example"
        version: "1.0"
        config:
          endpoint: "http://localhost:8080/services/echo"
          endpointType: "rest"
          public: true
          gateway: "test-gw"
        policies:
          - name: CachingPolicy
            config:
              ttl: 60
          - plugin: test-plugin
            config:
               Foo: 17
               Bar: "yes"
          - name: PolicyTwo # Can be elided if only one policyDef in plugin.
            plugin: another-test-plugin
            config:
               Foo: 17
               Bar: "yes"
```

**Comment by msavy** [comment]
Tests were passing and now fail despite nothing having changed (90fbd8c passed before, but now fails)

**Comment by msavy** [comment]
Hmm, I think CI might have broken

**Comment by msavy** [comment]
Pulling in @rpickard to give thoughts/review -- hope you don't mind.

**Comment by rpickard** [comment]
Hi Marc, am I right in thinking that this implementation will allow you to
push a set of api definitions directly to the gateway without going via the
manager?

**Comment by msavy** [comment]
@rpickard yes, exactly!

BTW, I'm doing some refactoring work on my local branch to make things work a bit more sensibly, as inheritance is causing some unnecessary duplication (turns out that args4j doesn't support Java 8 interface mixins for adding extra options).

**Comment by msavy** [comment]
Guys, I've done a bunch more work to move this over to JCommander (for reasons I'll go into more detail on) and think I should probably push onto a new PR or something.

**Comment by msavy** [comment]
See #28 instead, please.

**Pull request closed without merging**